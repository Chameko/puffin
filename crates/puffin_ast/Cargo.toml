[package]
name = "puffin_ast"
version = "0.0.1"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[lib]

[dependencies]
rowan = "0.15"