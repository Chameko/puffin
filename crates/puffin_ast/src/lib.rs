/// Describes all the syntax kinds in puffin
pub mod syntax_kind;
pub mod ast;
pub mod green;

use puffin_source::TextSlice;
use rowan::TextRange;
pub use syntax_kind::SyntaxKind;
pub use green::{SyntaxNode, SyntaxToken, SyntaxNodeChildren};
pub use ast::{AstPtr, SyntaxNodePtr, AstMap};

/// Converts between a [TextSlice] and a [TextRange]
pub fn text_range(slice: TextSlice) -> TextRange {
    TextRange::new((*slice.start()).into(), (*slice.end() + 1).into())
}
