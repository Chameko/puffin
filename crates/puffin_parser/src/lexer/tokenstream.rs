use puffin_ast::SyntaxKind;
use puffin_hir::source::TextSlice;

/// A token used in the lexer and parser
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Token {
    /// The type of the token
    pub ty: SyntaxKind,
    /// The column of line the token is on. The indexing starts from 0
    pub col: TextSlice,
}

impl Token {
    /// Creates a new token
    pub fn new(ty: SyntaxKind, col: TextSlice) -> Self {
        Self {
            ty,
            col,
        }
    }

    /// Extracts the tokens text from the provided source, split by lines.
    pub fn get_text<'a>(&self, src: &'a str) -> &'a str {
        src.get(self.col.clone()).expect("token out of range")
    }
}

/// A stream of tokens
#[derive(Debug, PartialEq, Eq, Clone)]
pub struct TokenStream {
    /// The internal tokens
    tokens: Vec<Token>,
    /// Which token we are currently at
    cursor: usize,
}

impl TokenStream {
    /// Create a new token stream
    pub fn new(tokens: Vec<Token>) -> Self {
        Self {
            tokens,
            cursor: 0,
        }
    }

    /// Peek at the next [Token]
    pub fn peek(&self) -> Option<&Token> {
        self.tokens.get(self.cursor + 1)
    }

    /// Get the next [Token]
    pub fn next(&mut self) -> Option<&Token> {
        self.cursor += 1;
        self.tokens.get(self.cursor)
    }

    /// Advance the cursor by one
    pub fn advance(&mut self) {
        self.cursor += 1;
    }

    /// Get the current [`Token`]. Note that current will always be the last valid token
    pub fn current(&self) -> Option<&Token> {
        self.tokens.get(self.cursor)
    }

    /// Gets the last token. This is primarily used for error generation. [`SyntaxKind::EOF`] is used if there are no tokens
    pub fn last(&self) -> Option<&Token> {
        self.tokens.last()
    }
}
