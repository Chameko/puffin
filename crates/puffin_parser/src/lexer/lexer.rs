use puffin_ast::SyntaxKind;
use std::{iter::Peekable, str::Chars};
use super::tokenstream::Token;

/// The lexer responsible for taking a raw file as a string and converting it to a flat array of [`Token`]
pub struct Lexer<'a> {
    /// The source the lexer is parsing
    src: Peekable<Chars<'a>>,
    /// Charaters that may form a keyword or identifier are stored here
    working: String,
    /// The current column
    col: u32,
}

impl<'a> Lexer<'a> {
    /// Next character
    #[inline]
    fn next(&mut self) -> Option<char> {
        self.src.next()
    }

    /// Peek one character ahead
    #[inline]
    fn peek(&mut self) -> Option<char> {
        self.src.peek().map(|c| *c)
    }

    ///  Creates a new lexer
    pub fn new(src: &'a str) -> Self {
        Self {
            src: src.chars().peekable(),
            working: String::new(),
            col: 0,
        }
    }

    /// Parses a number
    fn parse_number(s: &str) -> SyntaxKind {
        let mut ty = SyntaxKind::INT;
        for char in s.chars() {
            if !char.is_numeric() {
                if char == '.' {
                    ty = SyntaxKind::FLOAT;
                } else {
                    // Cannot have letters in a number
                    return ty;
                }
            }
        }
        ty
    }

    // Scans the working string and creates an identifier or a number
    fn scan_working(&mut self) -> Option<Token> {
        let ret = match self.working.as_str() {
            "and" => Some(SyntaxKind::KW_AND),
            "or" => Some(SyntaxKind::KW_OR),
            "print" => Some(SyntaxKind::KW_PRINT),
            "let" => Some(SyntaxKind::KW_LET),
            "fun" => Some(SyntaxKind::KW_FUN),
            "int" => Some(SyntaxKind::KW_INT),
            "float" => Some(SyntaxKind::KW_FLOAT),
            "bool" => Some(SyntaxKind::KW_BOOL),
            "string" => Some(SyntaxKind::KW_STRING),
            s => {
                if let Some(char) = s.chars().next() {
                    if char.is_numeric() {
                        Some(Self::parse_number(s))
                    } else {
                        Some(SyntaxKind::IDENT)
                    }
                } else {
                    None
                }
            }
        };
        let working_clone = std::mem::take(&mut self.working);
        if let Some(ty) = ret {
            let ret = Some(Token::new(ty, self.col..=(self.col + working_clone.len() as u32 - 1)));
            self.col += working_clone.len() as u32;
            ret
        } else {
            None
        }
    }

    /// Starts the scan of the file
    pub fn start_scan(mut self) -> Vec<Token> {
        let mut tokens = vec![];
        self.scan(&mut tokens);
        tokens
    }

    /// Used to record a symbol. Note: parses the working string before recording the symbol
    fn symbol(&mut self, ty: SyntaxKind, string: &str, tokens: &mut Vec<Token>) {
        if let Some(tk) = self.scan_working() {
            tokens.push(tk);
        }
        tokens.push(Token::new(ty, self.col..=(self.col + string.len() as u32 - 1)));
        self.col += string.len() as u32;
    }

    /// Scans the string into a flat array of [`SyntaxKind`] and [`String`]
    fn scan(&mut self, tokens: &mut Vec<Token>) {
        if let Some(char) = self.next() {
            match char {
                '&' => {
                    if let Some('&') = self.peek() {
                        self.next();
                        self.symbol( SyntaxKind::AMPAMP, "&&", tokens);
                    } else {
                        self.symbol( SyntaxKind::AMP, "&", tokens);
                    }
                }
                '|' => {
                    if let Some('|') = self.peek() {
                        self.next();
                        self.symbol( SyntaxKind::PIPEPIPE, "||", tokens);
                    } else {
                        self.symbol( SyntaxKind::PIPE, "|", tokens);
                    }
                }
                '=' => {
                    if let Some('=') = self.peek() {
                        self.next();
                        self.symbol( SyntaxKind::EQEQ, "==", tokens);
                    } else {
                        self.symbol( SyntaxKind::EQ, "=", tokens)
                    }
                }
                '>' => {
                    if let Some('=') = self.peek() {
                        self.next();
                        self.symbol( SyntaxKind::GTEQ, ">=", tokens);
                    } else {
                        self.symbol( SyntaxKind::GT, ">", tokens);
                    }
                }
                '<' => {
                    if let Some('=') = self.peek() {
                        self.next();
                        self.symbol( SyntaxKind::LTEQ, "<=", tokens);
                    } else {
                        self.symbol( SyntaxKind::LT, "<", tokens);
                    }
                }
                '.' => {
                    if let Some(c) = self.peek() {
                        // If the . is part of a number we push it
                        if c.is_numeric() {
                            self.working.push('.')
                        } else {
                            self.symbol( SyntaxKind::DOT, ".", tokens);
                        }
                    } else {
                        self.symbol( SyntaxKind::DOT, ".", tokens)
                    }
                }
                '/' => {
                    if let Some('/') = self.peek() {
                        self.next();
                        let mut peek = self.peek();
                        let mut comment = String::new();
                        while peek != Some('\n') && peek != None {
                            comment.push(self.next().unwrap());
                            peek = self.peek();
                        }
                        self.symbol( SyntaxKind::COMMENT, "//", tokens);
                    } else {
                        self.symbol( SyntaxKind::SLASH, "/", tokens);
                    }
                }
                ' ' => {
                    let mut length = 1;
                    let mut col = self.col;
                    while let Some(' ') = self.peek() {
                         length += 1;
                         self.next();
                    }
                    if let Some(tk) = self.scan_working() {
                        // Update the column as we may have scanned a new token
                        col = *tk.col.end() + 1;
                        tokens.push(tk);
                    }
                    tokens.push(Token::new(SyntaxKind::WHITESPACE, col..=(col + length - 1)));
                    self.col += length;
                }
                '+' => self.symbol( SyntaxKind::PLUS, "+", tokens),
                '-' => self.symbol( SyntaxKind::MINUS, "-", tokens),
                '*' => self.symbol( SyntaxKind::STAR, "*", tokens),
                '(' => self.symbol( SyntaxKind::L_PAREN, "(", tokens),
                ')' => self.symbol( SyntaxKind::R_PAREN, ")", tokens),
                '{' => self.symbol(SyntaxKind::L_BRACE, "{", tokens),
                '}' => self.symbol(SyntaxKind::R_BRACE, "}", tokens),
                '!' => self.symbol( SyntaxKind::EXCLAMATION, "!", tokens),
                ',' => self.symbol(SyntaxKind::COMMA, ",", tokens),
                ':' => self.symbol(SyntaxKind::COLON, ":", tokens),
                '\n' => self.symbol( SyntaxKind::NL, "\n", tokens),
                c => {
                    if c.is_alphanumeric() {
                        self.working.push(c)
                    } else {
                        self.symbol( SyntaxKind::ERROR, &c.to_string(), tokens)
                    }
                }
            }
            self.scan(tokens);
        } else {
            if let Some(tk) = self.scan_working() {
                tokens.push(tk);
            }
        }
    }
}

#[cfg(test)]
mod lexer_test {
    use super::Token;

    fn test_base(input: &str) -> String {
        let lexer = super::Lexer::new(input);
        let tokens = lexer.start_scan();
        tokens.into_iter().map(|t| {
            input.get(*t.col.start() as usize..=*t.col.end() as usize).unwrap()
        }).collect::<String>()
    }

    #[test]
    fn symbol_kinds() {
        let output = test_base("& | + - * / > < = ( ) ! . \n == >= <= && || { }");
        insta::assert_snapshot!(output);
    }

    #[test]
    fn keywords_and_ident() {
        let input = "and sometoiurng or skngeing";
        let lexer = super::Lexer::new(input);
        let tokens = lexer.start_scan();
        let output = tokens
            .into_iter()
            .map(|p| {
                match p {
                Token {ty: super::SyntaxKind::KW_AND, ..} => format!("|and:{}|", input.get(*p.col.start() as usize..=*p.col.end() as usize).unwrap()),
                Token {ty: super::SyntaxKind::KW_OR, ..} => format!("|or:{}|", input.get(*p.col.start() as usize..=*p.col.end() as usize).unwrap()),
                Token {ty: super::SyntaxKind::WHITESPACE, ..} => format!("|whitespace:{}|", input.get(*p.col.start() as usize..=*p.col.end() as usize).unwrap()),
                Token {ty: super::SyntaxKind::IDENT, ..} => format!("|ident:{}|", input.get(*p.col.start() as usize..=*p.col.end() as usize).unwrap()),
                Token {ty: super::SyntaxKind::SOURCE_FILE, ..}  => format!("|src:{}|", input.get(*p.col.start() as usize..=*p.col.end() as usize).unwrap()),
                _ => format!("error:{}", input.get(*p.col.start() as usize..=*p.col.end() as usize).unwrap()),
            }})
            .collect::<String>();
        insta::assert_snapshot!(output);
    }

    #[test]
    fn numbers() {
        let input = "1.23 42";
        let lexer = super::Lexer::new(input);
        let tokens = lexer.start_scan();
        let output = tokens
            .into_iter()
            .map(|p| match p {
                Token {ty: super::SyntaxKind::FLOAT, .. } => format!("|float:{}|", input.get(*p.col.start() as usize..=*p.col.end() as usize).unwrap()),
                Token {ty: super::SyntaxKind::INT, .. } => format!("|int:{}|", input.get(*p.col.start() as usize..=*p.col.end() as usize).unwrap()),
                Token {ty: super::SyntaxKind::WHITESPACE, .. } => format!("|whitespace:{}|", input.get(*p.col.start() as usize..=*p.col.end() as usize).unwrap()),
                Token {ty: super::SyntaxKind::SOURCE_FILE, .. }  => format!("|src:{}|", input.get(*p.col.start() as usize..=*p.col.end() as usize).unwrap()),
                _ => format!("error:{}", input.get(*p.col.start() as usize..=*p.col.end() as usize).unwrap()),
            })
            .collect::<String>();
        insta::assert_snapshot!(output);
    }
}
