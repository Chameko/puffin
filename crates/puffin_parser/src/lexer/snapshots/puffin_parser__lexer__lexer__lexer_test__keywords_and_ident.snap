---
source: crates/puffin_parser/src/lexer/lexer.rs
expression: output
---
|and:and||whitespace: ||ident:sometoiurng||whitespace: ||or:or||whitespace: ||ident:skngeing|
